### DISCUSSION TITLE
Exit on mobile platforms

### DISCUSSION THREAD

**Discussion by TojikCZ** [ppy/osu#16546] [Ideas] [answered]
Hi, yes, Android apps don't have exit buttons. But I can't hold my back button, because I don't have that button. I only got gestures and the only way to quit osu is using the app switcher. Please let me quit it with the "see you next time". I miss it :pleading_face:

**Comment by peppy** [comment, marked as answer]
You can't "quit" mobile applications.

**Reply by TojikCZ** [reply]
Do you mean, there's no Android API call to stop the app and remove it from recents? I haven't tested this, but the second part of this answer looks promising https://stackoverflow.com/a/27765687/2442813
Anyway, sorry for wasting your time. I might try to fiddle around with it. If I get something working, I'll bother you again.

**Reply by peppy** [reply]
It's anti-UX. Not going to happen.

**Reply by LittleEndu** [reply]
I don't know about iOS but on android it is very much the expected UX that the app will eventually quit to home screen if the back button is pressed enough. That is to say the app will keep going back to previous screen until there is no screen to go back to at which point it will exit.

**Reply by peppy** [reply]
It won't exit. It will background. Same as iOS.

**Reply by LittleEndu** [reply]
OK, but I don't think current release even goes to background. Or is it tied to "hold to confirm" and you can't hold down the back button unless you have a physical one?

**Reply by TojikCZ** [reply]
Imagine that the only way to quit Osu on Windows would be to alt tab. That's exactly what it's like on Android right now.

I wanted to gather more evidence before I came back but... GTA SA has an exit button, so that's worth something right?

But that's not what my request even was. I wanted the "see you next time" thing. That's the expected UX from Osu. You made it this way, 

Osu doesn't follow any other Android UX guidelines right? So why is this one so important?

**Reply by peppy** [reply]
We can make the back button return to the home screen for devices without a home button (does this exist?).

Exit flow will not be implemented on mobile. It is out of the question.

**Reply by LittleEndu** [reply]
Doubt that there would be devices without home buttons, except maybe ones that have their physical home button broken so it doesn't work anymore - that's essentially the same as no home button but you can't detect for it.

Also home button is quite jank on some manufacturers, where it is slow to react and there 50/50 chance that you will hail the assistant instead. At least that's why I prefer the back button to home or recent apps ones.

**Comment by Stedoss** [comment]
I am really struggling to see why this is not being considered, as this is a pretty standard Android behaviour. Clicking the back button on what is considered the home screen for the app should at least background the app and take the user back to their launcher, however it seems that this case is not handled currently (using the `back` gesture on the home screen currently does nothing, which is intentional).

I'm not a big mobile gamer, but it seems that it's quite standard to have a `Are you sure? Y/N` kind of modal that shows when you click back on the "home screen" of the game, then exits if confirmed. The two games I have on my device (Beatstar and the Microsoft Solitaire Collection) both handle a back button event on the home screen, prompting the user whether they would like to quit. This usually keeps the app showing in the `Recent Apps` page, but unloaded from memory until it is re-focused.

AnkiDroid does this, on the deck selector it prompts you when the back button is clicked, and that you can click the back button again to exit the app (this calls `finish()` on the main activity - https://github.com/ankidroid/Anki-Android/blob/cb15433b86c6b1a9a18fe9716d11354ac5c688ea/AnkiDroid/src/main/java/com/ichi2/anki/DeckPicker.java#L957).

**Reply by peppy** [reply]
Back button returning to home will be supported.

Unloading from memory is a given when the game becomes the non-active window. This is handled by the operating system.

**Reply by peppy** [reply]
Tracking at https://github.com/ppy/osu/issues/16899.
